### DISCUSSION TITLE
flask and tkinter for a dialog box (yes/no - start/stop server)

### DISCUSSION THREAD

**Discussion by Xavlight** [pallets/flask#4218] [Q&A]
Hi,

I would like to create a dialog box (tkinter) with an on/off button.
When the program starts, it starts the Flask server directly.
And with the button, you can stop or restart it.

But as soon as I run app.run, the dialog box does not work anymore.
I used parameter 'threaded=True', but not effect.
I tried with multithreading 'threading.Thread' to start flask and tkinter, it works. But in this case I can't stop the flask server on button click off.

Can anyone give me a hint?

Thanks

Environment:

Python version: 3.9.6
Flask version: 2.0.1
Tkinter : 8.6

**Comment by miguelgrinberg** [comment]
The Flask server is blocking, so you must start it on a background thread if you want your UI to remain operational.

To shutdown the server you have to send a request to it that calls the Werkzeug shutdown function. See the [documentation](https://werkzeug.palletsprojects.com/en/2.0.x/serving/#shutting-down-the-server) to learn how to do this.

**Comment by Yourun-proger** [comment]
Also, on [StackOverflow](https://stackoverflow.com/questions/15562446/how-to-stop-flask-application-without-using-ctrl-c) suggest using the following:
```py
from flask import request
def shutdown_server():
    func = request.environ.get('werkzeug.server.shutdown')
    if func is None:
        raise RuntimeError('Not running with the Werkzeug Server')
    func()
```
And in your GUI you can add something like this:
```py
btn = tk.Button(main_window, text='stop flask server', command=shutdown_server)
```

**Reply by ThiefMaster** [reply]
That's deprecated: https://github.com/pallets/werkzeug/pull/1873

**Reply by Yourun-proger** [reply]
Sorry for bad answer. I didn't know this
